### PR TITLE
Support 4 zero crossing dimmer RobotDyn

### PR DESCRIPTION
**Pull request by benoitLecuppe**
## Description:

**Related issue (if applicable):** fixes #<Tasmota issue number goes here>

## Checklist:
  - [x] The pull request is done against the latest development branch
  - [ ] Only relevant files were touched
  - [ ] Only one feature/fix was added per PR and the code change compiles without warnings
  - [ ] The code change is tested and works with Tasmota core ESP8266 V.2.7.4.9
  - [ ] The code change is tested and works with Tasmota core ESP32 V.2.0.3
  - [ ] I accept the [CLA](https://github.com/arendst/Tasmota/blob/development/CONTRIBUTING.md#contributor-license-agreement-cla).

_NOTE: The code change must pass CI tests. **Your PR cannot be merged unless tests pass**_

### CODE REVIEW

**Comment by barbudor** [comment]
Please name your PR properly and re-insert the PR submission template
Also remove your changes in `platformio.ini`.

File: tasmota/support_tasmota.ino [hunk 1]
```diff
@@ -203,8 +203,15 @@ void ZeroCrossMomentEnd(void) {
 
 void IRAM_ATTR ZeroCrossIsr(void) {
   uint32_t time = micros();
-  TasmotaGlobal.zc_interval = ((int32_t) (time - TasmotaGlobal.zc_time));
-  TasmotaGlobal.zc_time = time;
+  static bool lastValue;
+  bool value = digitalRead(Pin(GPIO_ZEROCROSS));
+  if(value && !lastValue) {
```
Comment (arendst) [inline tasmota/support_tasmota.ino:208]: Why this?

The ISR is called on a CHANGE of GPIO so this should ALWAYS be true.

File: tasmota/xdrv_62_robotdyn.ino [hunk 2]
```diff
@@ -0,0 +1,253 @@
...
+ * DIMMER4  GPIO12
+ * MCU ESP8266
+ * FLASH 4MB
+ * AC-IN 90-240V
+ * DC - VDD 3.3V
+ * RDB abreviation mean RoBotDyn
+\*********************************************************************************************/
+
+#define XDRV_62                 62
+
```
Comment (arendst) [inline tasmota/xdrv_62_robotdyn.ino:29]: XDRV_62 is already in use. Change to XDRV_63

File: tasmota/xdrv_62_robotdyn.ino [hunk 3]
```diff
@@ -0,0 +1,253 @@
...
+extern "C" int startWaveform(uint8_t pin, uint32_t timeHighUS, uint32_t timeLowUS, uint32_t runTimeUS = 0, int8_t alignPhase = -1, uint32_t phaseOffsetUS = 0, bool autoPwm = false);
+extern "C" int stopWaveform(uint8_t pin);
+
+
+void rbdInit(void){
+  // RobotDyn AC dimmer Init
+  AddLog(LOG_LEVEL_INFO, PSTR("========== RBD INIT rdbInit() "));
+  // init ZeroCross detection. One pin must be attach to "ZC Pulse"
+  if (!PinUsed(GPIO_ZEROCROSS))  AddLog(LOG_LEVEL_ERROR, PSTR("========== RBD INIT ERROR: on pin mist be set to ZC Pulse in template"));
```
Comment (arendst) [inline tasmota/xdrv_62_robotdyn.ino:38]: Change all AddLog(LOG_LEVEL_x, PSTR("====) functions to AddLog(LOG_LEVEL_x, PSTR("RBD: ) in line with Tasmota logging.

Also use text strings as short as possible, You're not writing a book! It should be cheatsheet.

File: tasmota/xdrv_62_robotdyn.ino [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+const char kRBDCommands[] PROGMEM = "|RBDTest" ;
+void (* const RBDCommands[])(void) PROGMEM = {&rbdCmndTest };
+// This function is ethe interface (the entry point for all action require by tastota) with tasmota program
+
+bool Xdrv62(uint8_t function) {
+
+  bool result = false;
+  switch(function) {
+    case FUNC_INIT : // module must init
+      rbdInit();
```
Comment (arendst) [inline tasmota/xdrv_62_robotdyn.ino:234]: A driver needs to handle it's back off when there it is not configured by either selecting a unique GPIO or a setting. 

Your driver, once enabled in the compile, will report an error if GPIO_ZEROCROSS is not defined on any device. Not user friendly. 

It should also backoff servicing any other function call if it's not initialized correctly.

File: tasmota/xdrv_63_robotdyn.ino [hunk 5]
```diff
@@ -0,0 +1,253 @@
+/*
+  xdrv_63_robotdyn.ino -support for robotdin 4 AC channel dimmer with zero crossing
+
+  SPDX-FileCopyrightText: 2022 Benoit Lecuppe
+
+  SPDX-License-Identifier: GPL-3.0-only
+*/
+//ifdef VSCODE_DEV
+  #define USE_LIGHT
```
Comment (ascillato) [inline tasmota/xdrv_63_robotdyn.ino:9]: This line shouldn't be here. Please, delete it.

File: tasmota/xdrv_63_robotdyn.ino [hunk 6]
```diff
@@ -0,0 +1,253 @@
...
+  #define USE_ROBOTDYN_DIMMER
```
Comment (ascillato) [inline tasmota/xdrv_63_robotdyn.ino:10]: This #define should go in my_user_config.h file and it should go commented out. Please, move it to my_user_config.h as:

```cpp
// #define USE_ROBOTDYN_DIMMER
```

Comment (ascillato) [inline tasmota/xdrv_63_robotdyn.ino:10]: This #define should go in my_user_config.h file and it should go commented out. Please, move it to my_user_config.h as:

```cpp
// #define USE_ROBOTDYN_DIMMER
```

**Review by ascillato — CHANGES_REQUESTED** [review]
Please, take a look to the comments. Thanks.

**Comment by ascillato** [comment]
The CI Tests are failing due to redefinition of

```cpp
  #define USE_LIGHT
  #define USE_ROBOTDYN_DIMMER
```

Please, take a look to the comments. Thanks.

**Comment by benoitLecuppe** [comment]
Hello everybody.
I'm sorry I didn't want to reintegrate my changes. it is a handling error. The program works well. but I'm not quite ready for reintegration
Benoit Lecuppe
French

**Pull request closed without merging**